Wallet txn activities

Jibz1 opened the pull request:
Wallet activities update. 
- Update  to match the latest design,
- I moved the from redux to react query
- Note (using the old icons)


https://user-images.githubusercontent.com/8676844/215287199-5e1d03fd-550b-41e0-a75b-4207d3bff043.mov


<img width="409" alt="Screenshot 2023-01-30 at 10 35 17 AM" src="https://user-images.githubusercontent.com/8676844/215531395-1bb12f2a-8231-435a-9fa9-d1fd0b9a306a.png">

<img width="436" alt="Screenshot 2023-01-31 at 10 04 45 AM" src="https://user-images.githubusercontent.com/8676844/215797180-8ce38727-278d-47d7-a058-e03e8dfff15c.png">

mystie711 commented:
The text should be center aligned vertically to the image thumb on left.
<img width="274" alt="image" src="https://user-images.githubusercontent.com/109102528/212803508-8135e69c-19c6-40ae-9718-7ea0d1c9ab4d.png">

No need to show "--" on the right when "amount" not available. This is true in general for "Call" and "Failed" transactions. We don't show gas fees in activity view

There should be ":" after To/From etc and a space
<img width="371" alt="image" src="https://user-images.githubusercontent.com/109102528/212804156-fba977b5-4c57-441b-adfc-539fb9914911.png">

Figma,
https://www.figma.com/file/xu2VnyPAuruNi74UQzGigW/02-Wallet?node-id=472%3A31082&t=AmxTK1eUwhs01oD2-1

Why does this "Sent" transaction have a "From" field? It should have a "To: 0x1234...5678" sort of info.
<img width="373" alt="image" src="https://user-images.githubusercontent.com/109102528/212804373-ba4974b5-350b-4e3b-8573-732da98b9620.png">

Jibz1 commented:
> The text should be center aligned vertically to the image thumb on left. <img alt="image" width="274" src="https://user-images.githubusercontent.com/109102528/212803508-8135e69c-19c6-40ae-9718-7ea0d1c9ab4d.png">
> 
> No need to show "--" on the right when "amount" not available. This is true in general for "Call" and "Failed" transactions. We don't show gas fees in activity view
> 
> There should be ":" after To/From etc and a space <img alt="image" width="371" src="https://user-images.githubusercontent.com/109102528/212804156-fba977b5-4c57-441b-adfc-539fb9914911.png">
> 
> Figma, https://www.figma.com/file/xu2VnyPAuruNi74UQzGigW/02-Wallet?node-id=472%3A31082&t=AmxTK1eUwhs01oD2-1
> 
> Why does this "Sent" transaction have a "From" field? It should have a "To: 0x1234...5678" sort of info. <img alt="image" width="373" src="https://user-images.githubusercontent.com/109102528/212804373-ba4974b5-350b-4e3b-8573-732da98b9620.png">

@mystie711 I have addressed those issues.

Jordan-Mysten commented on apps/wallet/src/ui/app/hooks/useGetTransaction.ts:
```diff
@@ -0,0 +1,54 @@
+    return useQuery(
+        ['transactions-by-address', normalizedAddress, network],
+        async () => {
+            const txnIds = await rpc.getTransactionsForAddress(
```
If we're rewriting this can we migrate to the new APIs that exist under `getTransactions`?

Jibz1 replied:
Done.

Jordan-Mysten commented on apps/wallet/src/ui/app/hooks/useGetTransaction.ts:
```diff
@@ -0,0 +1,54 @@
+    const rpc = useRpc();
+    return useQuery(
+        ['transaction-by-id', transactionId, network],
+        () => rpc.getTransactionWithEffects(transactionId),
```
I imagine you might want to batch these together? Seems like it'll be a lot of requests?

Jibz1 replied:
Done.

Jordan-Mysten commented on apps/wallet/src/ui/app/hooks/useGetTransaction.ts:
```diff
@@ -0,0 +1,54 @@
+        },
+        {
+            enabled: !!address,
+            refetchOnWindowFocus: true,
```
Isn't this the default?

Jordan-Mysten commented on apps/wallet/src/ui/app/hooks/useGetTransaction.ts:
```diff
@@ -0,0 +1,54 @@
+    const normalizedAddress = normalizeSuiAddress(address);
+    const rpc = useRpc();
+    return useQuery(
+        ['transactions-by-address', normalizedAddress, network],
```
Shouldn't need the network here, we should clear the query store when the network changes.

Jordan-Mysten commented on apps/wallet/src/ui/app/hooks/index.ts:
```diff
@@ -21,5 +21,9 @@ export { useTransactionDryRun } from './useTransactionDryRun';
+export {
+    useGetTransactionById,
+    useGetTransactionIdByAddress,
+} from './useGetTransaction';
```
Don't love filenames that imply exports but don't actually export the implied export. Would be fine if this was named `transactions`.

Jordan-Mysten commented on apps/wallet/src/ui/app/pages/home/transactions/index.tsx:
```diff
@@ -1,19 +1,55 @@
+        data: txns,
+        isError,
+        isLoading,
+    } = useGetTransactionIdByAddress(activeAddress || '');
```
seems like address should be optional instead of defaulting this

Jordan-Mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,235 @@
+const TRUNCATE_MAX_LENGTH = 8;
+const TRUNCATE_PREFIX_LENGTH = 4;
+
+function TxnItem({ txn }: { txn: SuiTransactionResponse }) {
```
I'll pull out into its own file.

mamos-mysten commented:
![Uploading image.png…]()

noticed we have two different arrow sizes

mamos-mysten commented:
![image](https://user-images.githubusercontent.com/122397493/213820974-9dcc98b6-0944-4175-b72e-3c1045839ca9.png)
text doesnt truncate with ellipsis as per design

mamos-mysten commented:
failed transaction doesnt have the correct icon/color
![image](https://user-images.githubusercontent.com/122397493/213821109-30be8eaa-3649-4a32-a365-2dc90dc18a89.png)

Jibz1 commented:
> ![image](https://user-images.githubusercontent.com/122397493/213820974-9dcc98b6-0944-4175-b72e-3c1045839ca9.png)
> text doesnt truncate with ellipsis as per design (there's an x-scrollbar)

That should not happen because [here](https://github.com/MystenLabs/sui/blob/fe3b7e791c6712e917f7a90db4a5e535ef4a457a/apps/wallet/src/ui/app/components/transactions-card/TxnImage.tsx#L12) both NFT meta are truncated

Jibz1 commented:
> failed transaction doesnt have the correct icon/color
> ![image](https://user-images.githubusercontent.com/122397493/213821109-30be8eaa-3649-4a32-a365-2dc90dc18a89.png)

Also, based on the txn response, I think the icon changes [here ](https://github.com/MystenLabs/sui/blob/fe3b7e791c6712e917f7a90db4a5e535ef4a457a/apps/wallet/src/ui/app/components/transactions-card/TxnIcon.tsx#L56)

mamos-mysten commented:
> 

this one may have been 

> > failed transaction doesnt have the correct icon/color
> > ![image](https://user-images.githubusercontent.com/122397493/213821109-30be8eaa-3649-4a32-a365-2dc90dc18a89.png)
> 
> Also, based on the txn response, I think the icon changes [here ](https://github.com/MystenLabs/sui/blob/fe3b7e791c6712e917f7a90db4a5e535ef4a457a/apps/wallet/src/ui/app/components/transactions-card/TxnIcon.tsx#L56)

looks like in the case of `txnFailed` we're using `SuiIcons.Info`

mamos-mysten commented:
> > ![image](https://user-images.githubusercontent.com/122397493/213820974-9dcc98b6-0944-4175-b72e-3c1045839ca9.png)
> > text doesnt truncate with ellipsis as per design (there's an x-scrollbar)
> 
> That should not happen because [here](https://github.com/MystenLabs/sui/blob/fe3b7e791c6712e917f7a90db4a5e535ef4a457a/apps/wallet/src/ui/app/components/transactions-card/TxnImage.tsx#L12) both NFT meta are truncated

ah, I was thinking this was handled by css so my test was invalid since it's happening in JS

Jibz1 commented:
This errors could be any number of things. It's best we display what it is.

Jibz1 commented:
> > > failed transaction doesnt have the correct icon/color
> > > ![image](https://user-images.githubusercontent.com/122397493/213821109-30be8eaa-3649-4a32-a365-2dc90dc18a89.png)
> > 
> > 
> > Also, based on the txn response, I think the icon changes [here ](https://github.com/MystenLabs/sui/blob/fe3b7e791c6712e917f7a90db4a5e535ef4a457a/apps/wallet/src/ui/app/components/transactions-card/TxnIcon.tsx#L56)
> 
> looks like in the case of `txnFailed` we're using `SuiIcons.Info` see [this line](https://github.com/MystenLabs/sui/blob/fe3b7e791c6712e917f7a90db4a5e535ef4a457a/apps/wallet/src/ui/app/components/transactions-card/TxnIcon.tsx#L58)
> 
> also @mystie711 called out that this should probably read `Insufficient Gas` - we should make sure any other potential error messages are surfaced in a friendly format

These errors could be any number of things. It's best we display what it is. the Fail icon is 
<img width="49" alt="Screenshot 2023-01-21 at 9 45 08 PM" src="https://user-images.githubusercontent.com/8676844/213898193-d873e534-c2b2-41ad-8572-1b924b98b595.png">

Jordan-Mysten commented on apps/wallet/src/ui/app/hooks/useGetTransactionsByAddress.ts:
```diff
@@ -0,0 +1,42 @@
+    return rpc.getTransactionWithEffectsBatch(txnsIds.data);
+}
+
+// Fetch transactions on mount and every 2 seconds
```
This isn't really accurate, it just is considered stale after 2 seconds, but that doesn't refetch. Honestly with 2 second stale time you might as well keep it at 0 (or increase it to something meaningful).

Jibz1 replied:
done

Jordan-Mysten commented on apps/wallet/src/ui/app/hooks/useGetTransactionsByAddress.ts:
```diff
@@ -0,0 +1,42 @@
+import { api } from '_redux/store/thunk-extras';
+const STALE_TIME = 1000 * 2;
+
+async function getTransactionsByAddress(
```
Why not just define this inline?

Jibz1 replied:
moved inline.

Jordan-Mysten commented on apps/wallet/src/ui/app/hooks/useGetTransactionsByAddress.ts:
```diff
@@ -0,0 +1,42 @@
+        ['transactions-by-address', address],
+        async () => {
+            if (!address) {
+                throw new Error('No wallet address provided');
```
Technically impossible so you could also just non-null assert but I'm also totally fine keeping this here.

Jordan-Mysten commented on apps/wallet/src/shared/utils/index.ts:
```diff
@@ -23,17 +23,13 @@ export function usePageView() {
         customRPC && apiEnv === 'customRPC' ? customRPC : apiEnv.toUpperCase();
 
     useEffect(() => {
-        // NOTE: This is a hack to work around hook timing issues with the Growthbook SDK.
```
Bug doesn't look like it's fixed??

Jibz1 replied:
this is a mistake.

Jordan-Mysten commented on apps/wallet/src/ui/app/pages/home/transactions/index.tsx:
```diff
@@ -1,19 +1,71 @@
+    const selectedApiEnv = useAppSelector(({ app }) => app.apiEnv);
+    const queryClient = useQueryClient();
+
+    // Switching between accounts should invalidate the query cache
```
This seems unnecessary, just leave it in cache and let it get evicted at some point.

Jibz1 replied:
removing this

Jordan-Mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,216 @@
+            certificate.data.transactions[0]
+        )?.objectRef?.objectId;
+        return transferId
+            ? [transferId]
```
Why construct this into an array only to only use the first value from it?

Jordan-Mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,216 @@
+        address || ''
+    );
+
+    const objectIds = useMemo(() => {
```
Doesn't really need to be memo-d

Jibz1 replied:
done

Jordan-Mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,216 @@
+                                </div>
+                                <div className="flex flex-col w-full">
+                                    <div className="flex flex-col w-full gap-1.5">
+                                        <TxnTypeLabel
```
Why not do the label computation in this component instead of hoisting this logic to the parent component?

Jibz1 replied:
done

Jordan-Mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,216 @@
+        return recipientAddress;
+    }, [certificate.data.transactions, recipientAddress, txnKind]);
+
+    const isMint = txnKind === 'Call' && moveCallTxn?.function === 'mint';
```
Can we stop special casing mint function transactions please?

Jordan-Mysten commented on apps/wallet/src/ui/app/components/transactions-card/TxnImage.tsx:
```diff
@@ -0,0 +1,45 @@
+//TODO merge all NFT image displays
+export function TxnImage({ id }: { id: string }) {
+    const nftMeta = useGetNFTMeta(id);
+    const name = useMiddleEllipsis(
```
I feel like truncation should happen on the end not the middle right? Makes more sense to me to lean into CSS for this instead of using middle ellipsis truncation.

Jibz1 replied:
done

Jordan-Mysten commented on apps/wallet/src/ui/app/components/transactions-card/TxnIcon.tsx:
```diff
@@ -0,0 +1,66 @@
+        return isSender ? 'Send' : 'Receive';
+    }, [isSender, txnKindName]);
+
+    const icons = {
```
You can hoist this out of the component.

Jibz1 replied:
done

mystie711 reviewed (approved):
Core layout update looks great.

As discussed, will capture the Staking flow related UI updates in a JIRA once i have the design documented in Figma.

mamos-mysten commented on apps/wallet/src/ui/app/pages/home/transactions/index.tsx:
```diff
@@ -1,19 +1,61 @@
+            <div className="p-2">
+                <Alert mode="warning">
+                    <div className="mb-1 font-semibold">
+                        Something went wrong
```
small nit - `mb-1` class here causes the text in the alert banner to be off-center

<img width="329" alt="image" src="https://user-images.githubusercontent.com/122397493/214238775-dedf81ba-f9e8-47f1-bd64-faf77e03c945.png">

mamos-mysten replied:
can also drop the container div and pass the `p-2` class into `<Alert />`

Jibz1 replied:
done

mamos-mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,199 @@
+    address: SuiAddress;
+}) {
+    const { certificate } = txn;
+    const executionStatus = getExecutionStatusType(txn) as 'Success' | 'Failed';
```
it looks like this returns `success` or `failure` not sure why we're casting this value

Jibz1 replied:
you are right, removing

mamos-mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,199 @@
+            <div className="flex items-start w-full justify-between gap-3">
+                <div className="w-7.5">
+                    <TxnIcon
+                        txnFailed={executionStatus === 'Failed' || !!error}
```
I don't think `executionStatus` can ever `=== 'Failed'` based on the return value of `getExecutionStatus`

Jibz1 replied:
removed it

mamos-mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,199 @@
+    }, [txn]);
+
+    const error = useMemo(() => getExecutionStatusError(txn), [txn]);
+    const isSuTransfer =
```
`isSuiTransfer`

mamos-mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,204 @@
+    const txnKind = getTransactionKindName(certificate.data.transactions[0]);
+    const { coins: eventsSummary } = getEventsSummary(
+        txn.effects,
+        address || ''
```
these `address || ''` aren't needed based on https://github.com/MystenLabs/sui/pull/7437/files#diff-63e205db9d65086ba15d3650fca3f854442d7ffd4c2cac04efdff40f9e12d7b5R47

mamos-mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,204 @@
+        )?.objectRef?.objectId;
+        return transferId
+            ? transferId
+            : getTxnEffectsEventID(txn.effects, address || '')[0];
```
same here

mamos-mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,204 @@
+    );
+
+    const amount = useMemo(() => {
+        const amount = amountByRecipient && amountByRecipient[0]?.amount;
```
nit - could be `amountByRecipient?.[0]?.amount`

mamos-mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,204 @@
+                                Transaction Failed
+                            </Text>
+                            <div className="p-2">
+                                <Alert mode="warning">
```
dont know if we want to use Alert here based on the design

<img width="362" alt="image" src="https://user-images.githubusercontent.com/122397493/214244280-f0386cd2-611b-405b-82ef-e25a5a4f20ac.png">

Jibz1 replied:
reverted

mamos-mysten commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,205 @@
+        return Math.abs(amount || amountTransfers);
+    }, [amountByRecipient, eventsSummary]);
+
+    const recipientAddress = useMemo(() => {
```
I think ideally we'd want to move all this derivation logic out of the component into re-reusable hooks that could be composed into something for this view. especially if we're going to re-use this logic in other parts of the app, especially if we're going to need to use this logic elsewhere

Jibz1 replied:
I am not sure where else we will use this

pchrysochoidis commented on apps/wallet/src/ui/app/hooks/useGetTxnRecipientAddress.ts:
```diff
@@ -0,0 +1,43 @@
+
+export function useGetTxnRecipientAddress({ txn, address }: Props) {
+    const { certificate } = txn;
+    const { coins: eventsSummary } = getEventsSummary(txn.effects, address);
```
I think we should memo this one as well, otherwise  eventsSummary will always be a new array

pchrysochoidis reviewed (commented):
not sure about the logic in Transaction.tsx but rather than that looks ok to me

pchrysochoidis commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,198 @@
+        txnKind === 'PaySui' ||
+        txnKind === 'TransferSui' ||
+        txnKind === 'PayAllSui' ||
+        txnKind === 'Pay';
```
Pay I think is not only sui

Jibz1 commented:
> not sure about the logic in Transaction.tsx but rather than that looks ok to me
> 
> 
> 
> are we moving away from the RecentTransaction component. Can we clean up afterwards and remove the unused code?

Yes, that is the idea.

pchrysochoidis commented on apps/wallet/src/ui/app/components/transactions-card/Transaction.tsx:
```diff
@@ -0,0 +1,198 @@
+        txn.effects
+    );
+
+    const transferAmount = useMemo(() => {
```
One transaction can include multiple coins (of different type) moved from and to an account. Maybe we should start showing that to the UI somehow instead of showing always 1 or nothing.

Jibz1 replied:
done

mamos-mysten commented:
is there a design for an empty activity feed? 

![image](https://user-images.githubusercontent.com/122397493/215901539-af3160c6-b883-4abf-bc54-902718be5c24.png)

cc: @Jibz1 @mystie711

mamos-mysten commented:
just a callout - we might want to handle the case that image fails to load via `ipfs.io` ... in this case, it was a `429 Too Many Requests`

![image](https://user-images.githubusercontent.com/122397493/215904337-4d75732f-7b14-4c64-ab6e-2846285fda96.png)

cc: @mystie711

Jibz1 commented:
That is wrong we use ipfs

mystie711 commented:
> is there a design for an empty activity feed?
> 
> ![image](https://user-images.githubusercontent.com/122397493/215901539-af3160c6-b883-4abf-bc54-902718be5c24.png)
> 
> cc: @Jibz1 @mystie711

It did not exist before but i created this one for a quick fix...if so.
https://www.figma.com/file/xu2VnyPAuruNi74UQzGigW/02-Wallet?node-id=5207%3A84492&t=eBTEr73RFtc4HWpl-1

mystie711 commented:
> just a callout - we might want to handle the case that image fails to load via `ipfs.io` ... in this case, it was a `429 Too Many Requests`
> 
> ![image](https://user-images.githubusercontent.com/122397493/215904337-4d75732f-7b14-4c64-ab6e-2846285fda96.png)
> 
> cc: @mystie711

We are supposed to show the "placeholder image" thumbnail in such cases. The same solution we do for NFTs tab when the NFT images fails to load. Could we do that @Jibz1

pchrysochoidis commented on apps/wallet/src/ui/app/pages/home/transactions/index.tsx:
```diff
@@ -1,19 +1,79 @@
+        isLoading,
+        isError,
+        error,
+    } = useQuery<SuiTransactionResponse[], Error>(
```
Actually might be a good idea to move this one to it's own file as a hook
